### PR TITLE
Anonymization

### PR DESCRIPTION
**Pull request by na-g**
Adds the ability to anonymize the parsed SDP by masking sensitive information.

### CODE REVIEW

**Comment by na-g** [comment]
Updated to remove clippy directive and integrate assert_eq! changes.

**Comment by na-g** [comment]
Looks like I used an interface that is unstable in 1.30.

File: src/lib.rs [hunk 1]
```diff
@@ -66,6 +69,15 @@ impl ToString for SdpConnection {
     }
 }
 
+impl AnonymizingClone for SdpConnection {
+    fn masked_clone(&self, anon: &mut StatefulSdpAnonymizer) -> Self {
+        let mut masked = self.clone();
+        masked.addr = anon.mask_ip(&self.addr);
```
Comment (nils-ohlmeier) [inline src/lib.rs]: I'm not sure I like that this first clone's the whole thing, to then free the cloned addr and replace it with the masked.
Wouldn't it make more sense to invoke the SdpConnection constructor instead? The arguments for the constructor would be direct clones of self attributes which don't require masking. And invoke the masking functions on the ones which require masking.
Just trying to reduce the cloning here.

File: src/anonymizer.rs [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+
+impl ToBytesVec for u64 {
+    fn to_byte_vec(&self) -> Vec<u8> {
+        let mut bytes = Vec::new();
+        let mut val = *self;
+        for _ in 0..8 {
+            bytes.push(val as u8);
+            val <<= 8;
+        }
+        bytes.reverse();
```
Comment (nils-ohlmeier) [inline src/anonymizer.rs]: Can't you just omit the semicolon on this line and delete the next line?

File: src/attribute_type.rs [hunk 3]
```diff
@@ -295,6 +308,15 @@ impl ToString for SdpAttributeRemoteCandidate {
     }
 }
 
+impl AnonymizingClone for SdpAttributeRemoteCandidate {
+    fn masked_clone(&self, anon: &mut StatefulSdpAnonymizer) -> Self {
+        let mut masked = self.clone();
```
Comment (nils-ohlmeier) [inline src/attribute_type.rs]: I think the only value you need to clone without masking it is component in here. Thus I would suggest to invoke the constructor instead.

File: src/attribute_type.rs [hunk 4]
```diff
@@ -634,6 +656,14 @@ impl ToString for SdpAttributeFingerprint {
     }
 }
 
+impl AnonymizingClone for SdpAttributeFingerprint {
+    fn masked_clone(&self, anon: &mut StatefulSdpAnonymizer) -> Self {
+        let mut masked = self.clone();
```
Comment (nils-ohlmeier) [inline src/attribute_type.rs]: With only two attributes I would prefer the constructor here over cloning.

File: src/anonymizer.rs (hunk 2 above)
Comment (na-g) [reply src/anonymizer.rs]: reverse() doesn't return Self. Oh, for Ruby's `tap`.

**Pull request merged**